File: mps_matrix_multiply.m
#include "mps_matrix_multiply.h"

void matrixMultiplyOnDeviceWithOffset(
    void *deviceID,
    void *commandBufferID,

    void *matrixAID,
    void *matrixBID,
    void *matrixCID,

    int _interiorColumns,
    float _alpha, float _beta,
    bool _transposeLeft, bool _transposeRight
) {
    id<MTLDevice> device = (id<MTLDevice>)deviceID;
    id<MTLCommandBuffer> commandBuffer = (id<MTLCommandBuffer>)commandBufferID;

    MPSMatrix *matrixA = (__bridge MPSMatrix*)matrixAID;
    MPSMatrix *matrixB = (__bridge MPSMatrix*)matrixBID;
    MPSMatrix *matrixC = (__bridge MPSMatrix*)matrixCID;

    MPSMatrixMultiplication *kernel = [[MPSMatrixMultiplication alloc]
        initWithDevice:device
        transposeLeft:_transposeLeft
        transposeRight:_transposeRight
        resultRows:matrixC.rows
        resultColumns:matrixC.columns
        interiorColumns:_interiorColumns
        alpha:_alpha
        beta:_beta];

    [kernel encodeToCommandBuffer:commandBuffer leftMatrix:matrixA rightMatrix:matrixB resultMatrix:matrixC];
    [kernel release];
}

File: mps_matrix_multiply.go
package mps

func (b *MTLCommandBuffer) MatrixMultiplyAB(aM, bM, cM *Matrix, alpha, beta float32) {
	b.MatrixMultiply(aM, bM, cM, aM.cols, false, false, alpha, beta)
}

func (b *MTLCommandBuffer) MatrixMultiplyATB(aM, bM, cM *Matrix, alpha, beta float32) {
	b.MatrixMultiply(aM, bM, cM, aM.cols, false, true, alpha, beta)
}

func (b *MTLCommandBuffer) MatrixMultiplyTAB(aM, bM, cM *Matrix, alpha, beta float32) {
	b.MatrixMultiply(aM, bM, cM, bM.rows, true, false, alpha, beta)
}

func (b *MTLCommandBuffer) MatrixMultiply(
	aM, bM, cM *Matrix,
	iC int,
	aT, bT bool,
	alpha, beta float32,
) {
	b.mu.Lock()
	defer b.mu.Unlock()
	b.uncommitted++

	mpsMatrixMultiply(b.deviceID, b.id, aM.matrixID, bM.matrixID, cM.matrixID, iC, aT, bT, alpha, beta)
}
